tests/test/util/test_scheduling.cpp

```diff
@@ -93,4 +95,283 @@ TEST_CASE("Test converting point-to-point mappings to scheduling decisions",
+                 "[util][scheduling-decision]")
+{
+    SchedulingConfig config = {
+        .hosts = { masterHost, "hostA" },
+        .slots = { 1, 1 },
+        .numReqs = 3,
+        .forceLocal = false,
+        .topologyHint = SchedulingTopologyHint::NORMAL,
+        .expectedHosts = { masterHost, "hostA", masterHost },
+    };
+
+    auto req = batchExecFactory("foo", "bar", config.numReqs);
+
+    testActualSchedulingDecision(req, config);
+}
+
+TEST_CASE_METHOD(SchedulingDecisionTestFixture,
+                 "Test force local forces executing at master",
+                 "[util][scheduling-decision]")
+{
+    SchedulingConfig config = {
+        .hosts = { masterHost, "hostA" },
+        .slots = { 1, 1 },
+        .numReqs = 2,
+        .forceLocal = false,
+        .topologyHint = SchedulingTopologyHint::NORMAL,
+        .expectedHosts = { masterHost, "hostA" },
+    };
+
+    auto req = batchExecFactory("foo", "bar", config.numReqs);
+
+    SECTION("Force local off")
+    {
+        config.forceLocal = false;
+        config.expectedHosts = { masterHost, "hostA" };
+    }
+
+    SECTION("Force local on")
+    {
+        config.forceLocal = true;
+        config.expectedHosts = { masterHost, masterHost };
+    }
+
+    testActualSchedulingDecision(req, config);
+}
+
+TEST_CASE_METHOD(SchedulingDecisionTestFixture,
+                 "Test master running out of resources",
+                 "[util][scheduling-decision]")
+{
+    SchedulingConfig config = {
+        .hosts = { masterHost, "hostA" },
+        .slots = { 0, 2 },
+        .numReqs = 2,
+        .forceLocal = false,
+        .topologyHint = SchedulingTopologyHint::NORMAL,
+        .expectedHosts = { "hostA", "hostA" },
+    };
+
+    auto req = batchExecFactory("foo", "bar", config.numReqs);
+
+    testActualSchedulingDecision(req, config);
+}
+
+TEST_CASE_METHOD(SchedulingDecisionTestFixture,
+                 "Test scheduling decision with many requests",
+                 "[util][scheduling-decision]")
+{
+    SchedulingConfig config = {
+        .hosts = { masterHost, "hostA", "hostB", "hostC" },
+        .slots = { 0, 0, 0, 0 },
+        .numReqs = 8,
+        .forceLocal = false,
+        .topologyHint = SchedulingTopologyHint::NORMAL,
+        .expectedHosts = { masterHost, masterHost, masterHost, masterHost },
+    };
+
+    auto req = batchExecFactory("foo", "bar", config.numReqs);
+
+    SECTION("Even slot distribution across hosts")
+    {
+        config.slots = { 2, 2, 2, 2 };
+        config.expectedHosts = { masterHost, masterHost, "hostA", "hostA",
+                                 "hostB",    "hostB",    "hostC", "hostC" };
+    }
+
+    SECTION("Uneven slot ditribution across hosts")
+    {
+        config.slots = { 3, 2, 2, 1 };
+        config.expectedHosts = { masterHost, masterHost, masterHost, "hostA",
+                                 "hostA",    "hostB",    "hostB",    "hostC" };
+    }
```

Comment:
The way these tests are parameterised is great, really nice and clear. We could probably make more use of use-case-specific fixtures like this to make the tests more readable and modifiable.